### PR TITLE
[JENKINS-14789] Configurable interval for computer retention check

### PR DESCRIPTION
**Pull request by das7pad**
See [JENKINS-14789](https://issues.jenkins.io/browse/JENKINS-14789).
Split from #5725 

The new config entry has been placed below the configuration of executors on the jenkins master.

image

I've considered putting the new option under the `/manage/computer/` namespace, but there is no existing configuration (page) that can be extended for this purpose. The config file for `/manage/computer/configure` is `/var/lib/jenkins/nodeMonitors.xml`  and it contains a list of `hudson.node_monitors.NodeMonitor` classes.

Reconfiguration at run-time is possible. Cancellation of the previous cycle is done on a best effort basis by cancelling the previous timer. (Awaiting the completion of old task throws an `IllegalMonitorStateException` and synchronizing is not possible as the task variable is not final. Concurrent runs are OK, so I did not bother ensuring no concurrency.) See my inline comment on why concurrent runs are OK:

>     * Restart the periodic check.
>     * The previous run may still be pending. We do not need to wait for its completion as all calls of
>     * {@link RetentionStrategy#check} inside the runner are performed in a queue with a lock.
>     * Per the docs on {@link RetentionStrategy#check}, it is OK to recheck earlier or later than requested.

---

Also update the lower bound delay as returned by each RetentionStrategy
 from 1min to 0min. This is enabling check intervals of <1min to trigger
 a re-check in the next cycle (previously, the nextCheck tracking would
 defer the next re-check into the 2nd cycle -- one for the timer loop
 and once for the nextCheck timestamp).

Includes a bug fix (from the last PR):
`CloudSlaveRetentionStrategy` tracks the agent timeout in milliseconds
 internally and returned the nextCheck delay as a tenth of that.
The nextCheck delay is consumed in minutes, leading to the max delay
 of one hour being used.
Fix: Convert the delay from milliseconds to minutes.

### Proposed changelog entries

- Entry 1: 14789, The latency for bringing up offline agents can be improved using a new global config option "Computer Retention Check Interval" and setting an "In demand delay" of zero on the agents.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change) and are in the imperative mood. [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  - Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
      _Any suggestions on adding auto-tests for the retention strategies? (current there are none to extend)_
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadoc, as appropriate.
     _Does this apply to models as well?_
- [x] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")` if applicable.
- [x] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content-Security-Policy directives (see [documentation on jenkins.io](https://www.jenkins.io/doc/developer/security/csp/)).
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Maintainer checklist

Before the changes are marked as `ready-for-merge`:

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are accurate, human-readable, and in the imperative mood
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/7037"></a>

### CODE REVIEW

File: core/src/main/java/hudson/slaves/ComputerRetentionWork.java
Comment (NotMyFault) [inline core/src/main/java/hudson/slaves/ComputerRetentionWork.java:66]:
```suggestion
     * @since TODO
     */
```

If this is intended to be not restricted and therefore public API, please add a since annotation.

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (NotMyFault) [inline core/src/main/java/jenkins/model/Jenkins.java:2383]: Please add javadocs as appropriate.

File: core/src/main/resources/jenkins/model/GlobalComputerRetentionCheckIntervalConfiguration/config.groovy
Comment (NotMyFault) [inline core/src/main/resources/jenkins/model/GlobalComputerRetentionCheckIntervalConfiguration/config.groovy:1]: Please use Jelly for newly added UI files.

File: core/src/main/java/hudson/slaves/ComputerRetentionWork.java
Comment (das7pad) [reply core/src/main/java/hudson/slaves/ComputerRetentionWork.java:68]: I've added the since annotation via 0fefa24987f2ba752eb793a37bf554c0ea03e5d8.

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (das7pad) [reply core/src/main/java/jenkins/model/Jenkins.java:2388]: Done via 0fefa24987f2ba752eb793a37bf554c0ea03e5d8.

File: core/src/main/resources/jenkins/model/GlobalComputerRetentionCheckIntervalConfiguration/config.groovy
Comment (das7pad) [reply core/src/main/resources/jenkins/model/GlobalComputerRetentionCheckIntervalConfiguration/config.groovy]: Done via 0fefa24987f2ba752eb793a37bf554c0ea03e5d8 and re-tested. I've seen two variants of providing the help content. For variant a) all the content is placed in a regular HTML file in `war/src/main/webapp/help` and for variant b) a stub jelly file refers to an entry in a dedicated resource bundle with a single key. The code for specifying the actual content with variant b) is [not easy to read](https://github.com/jenkinsci/jenkins/blob/9895d9f1b9b64860dd854639a9cc0cfbabe15fd5/core/src/main/resources/hudson/slaves/RetentionStrategy/Demand/help.properties#L1-L0) (e.g. to spot issues with the HTML markup) as IDEs do not recognize the strings as HTML for highlighting, so I went with a).

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (daniel-beck) [inline core/src/main/java/jenkins/model/Jenkins.java:2400]: Does not prevent submissions with the value 0.

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (daniel-beck) [inline core/src/main/java/jenkins/model/Jenkins.java:591]: I _think_ Jenkins has a special deserialization behavior, but it'd be safer if there's a test asserting that an existing Jenkins home will not set this to 0 on startup.

File: core/src/main/java/hudson/model/PeriodicWork.java
Comment (daniel-beck) [inline core/src/main/java/hudson/model/PeriodicWork.java:114]: Needs Javadoc and `@since`.

**Review by daniel-beck — DISMISSED** [review]
Please add test coverage.

Please explain why `ComputerRetentionWork` is not changed to an `AperiodicWork` instead.

**Comment by basil** [comment]
> Please explain why `ComputerRetentionWork` is not changed to an `AperiodicWork` instead.

Hello @daniel-beck, for my own education, can you please provide the reasoning behind this request? Providing the reasoning when making requests of others helps to facilitate a rational discussion. Thanks, Basil

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (daniel-beck) [inline core/src/main/java/jenkins/model/Jenkins.java:591]: Why is this added to `Jenkins`, instead of the new `GlobalConfiguration`?

**Comment by daniel-beck** [comment]
> Hello @daniel-beck, for my own education, can you please provide the reasoning behind this request? Providing the reasoning when making requests of others helps to facilitate a rational discussion. Thanks, Basil

The Javadoc of `AperiodicWork` seems to match pretty much exactly what you'd need for a reconfigurable interval:

> Extension point which allows scheduling a task with variable interval. Interval in evaluated every time before next task is scheduled by calling [getRecurrencePeriod()](https://javadoc.jenkins.io/hudson/model/AperiodicWork.html#getRecurrencePeriod()). Task to be scheduled is obtain by calling [getNewInstance()](https://javadoc.jenkins.io/hudson/model/AperiodicWork.html#getNewInstance()).
> 
> This class is similar to [PeriodicWork](https://javadoc.jenkins.io/hudson/model/PeriodicWork.html). The main difference is in re-evaluating delay interval every time. See [PeriodicWork](https://javadoc.jenkins.io/hudson/model/PeriodicWork.html) for details. Analog of [AsyncPeriodicWork](https://javadoc.jenkins.io/hudson/model/AsyncPeriodicWork.html) is [AsyncAperiodicWork](https://javadoc.jenkins.io/hudson/model/AsyncAperiodicWork.html).

So it is unclear to me why this approach was chosen. No explanation is provided.

**Comment by basil** [comment]
> The Javadoc of `AperiodicWork` seems to match pretty much exactly what you'd need for a reconfigurable interval

Thank you Daniel, this was not obvious to me and I suspect it was not obvious to the author of this PR either.

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (basil) [reply core/src/main/java/jenkins/model/Jenkins.java:591]: Why should it be added to the new `GlobalConfiguration`?

Comment (daniel-beck) [reply core/src/main/java/jenkins/model/Jenkins.java:591]: Ideally we'd not make Jenkins even more of a god object. This seems like a reasonable option to store elsewhere, not unlike `JenkinsLocationConfiguration`.

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (das7pad) [reply core/src/main/java/jenkins/model/Jenkins.java]: Fixed as part of moving the variable into the model dab113783ecfdf84685bcf8ead4c178c5fdc6c5f.

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (das7pad) [reply core/src/main/java/jenkins/model/Jenkins.java]: Done via c962e36afa6e6af051d8cfde2436424019fcb65f. I've also added a 60s fallback to the getter when the serialized value is not greater than zero.

File: core/src/main/java/hudson/model/PeriodicWork.java
Comment (das7pad) [reply core/src/main/java/hudson/model/PeriodicWork.java:119]: Done via 7faed41d1df856493cd3157ce79db98ca2521485

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (das7pad) [reply core/src/main/java/jenkins/model/Jenkins.java]: I adapted the storage from the `quietPeriod`/`scmRetryCount` config options, which both store their value in the `Jenkins` instance.
I moved the value into the model via dab113783ecfdf84685bcf8ead4c178c5fdc6c5f and repeated manual testing for (de-)serialization/updating via the GUI.

**Comment by das7pad** [comment]
Hello @daniel-beck,

Thank you for your feedback!

I've added tests for the model on (de-)serialization, checking of bounds and
 handling of default values via c962e36afa6e6af051d8cfde2436424019fcb65f.
Do they cover your concerns on deserialization?

Do you have any suggestions for testing the other changes (if necessary)?

I was not aware of `AperiodicWork`, thanks for the pointer.
`AperiodicWork` looks like a good fit for this use case -- on first glance.
On a second glance, it does not handle updating of the interval well.
The user experience is poor when decreasing the interval, say from 24h to 10s.
It can take hours for the previous timer (with 24h delay) to complete and
 only after that happened, it will eventually change to the new interval (10s).
I would consider this delayed propagation of the new interval a bug.
What are your thoughts on the updating behavior?

**Comment by daniel-beck** [comment]
>  The user experience is poor when decreasing the interval, say from 24h to 10s.
> It can take hours for the previous timer (with 24h delay) to complete and
> only after that happened, it will eventually change to the new interval (10s).
> I would consider this delayed propagation of the new interval a bug.
> What are your thoughts on the updating behavior?

That would be pretty bad, true, but what's a range we actually expect users to enter? Previously it was once a minute, IIUC, and waiting for up to a minute (an average of 30 seconds) for the interval to be set to a few seconds seems reasonable.

**Comment by das7pad** [comment]
> > The user experience is poor when decreasing the interval, say from 24h to 10s.
> > It can take hours for the previous timer (with 24h delay) to complete and
> > only after that happened, it will eventually change to the new interval (10s).
> > I would consider this delayed propagation of the new interval a bug.
> > What are your thoughts on the updating behavior?
> 
> That would be pretty bad, true, but what's a range we actually expect users to enter? Previously it was once a minute, IIUC, and waiting for up to a minute (an average of 30 seconds) for the interval to be set to a few seconds seems reasonable.

Good point. For the start we can limit the interval to 1-60s (38378cf0690c17b34178e571d4dd8ce2af276da0) and use the `AperiodicWork` scheduling (5a25749b8f2316a5c06fd7b26b7983b14b034ae0).

**Review by Kevin-CB — APPROVED** [review]
Looks good from a security aspect!

**Comment by das7pad** [comment]
Hi, is there anything I can help with to get this PR over the line?

**Comment by daniel-beck** [comment]
@das7pad I apologize for taking so long to remove my request for change after you addressed the issue. Queue and clouds is not something I'm very familiar with so my feedback is limited to more general Jenkins API use and I'm unlikely to be able to approve this PR on my own.

**Comment by das7pad** [comment]
No worries, @daniel-beck :)

Can you recommend any other members for review of this PR?

**Comment by das7pad** [comment]
Hi! Is there anything I can help with to bring this PR over the line?

**Comment by das7pad** [comment]
Hi! Is there anything I can help with to get this PR over the line?

FWIW: I'm running a small Jenkins instance with this PR for upcoming 2 years. It's been rock stable, bringing up ssh-agents reliably within a few seconds after being in need (In demand delay = 0 minutes set on agents, Computer Retention Check Interval = 5 seconds set globally). There is no noticeable increase in CPU usage, the check-loop runs on about 10 nodes. This is greatly cutting down the latency for getting my builds scheduled on an idle node by bringing up new capacity right away.

cc @timja @oleg-nenashev

**Review by timja — COMMENTED** [review]
I'm not that familiar with this area but I think ok, a couple of minor improvements

File: core/src/main/resources/jenkins/model/GlobalComputerRetentionCheckIntervalConfiguration/config.jelly
Comment (das7pad) [reply core/src/main/resources/jenkins/model/GlobalComputerRetentionCheckIntervalConfiguration/config.jelly]: This snippet has been adapted from another global setting (back in 2022, things might have changed since). Your suggestion works for me :+1:

**Comment by das7pad** [comment]
Thanks for the feedback @timja . I've pushed 26813f25e640726ab8ea7dc0c089d46ead359c63 to address your points.

File: core/src/main/resources/jenkins/model/GlobalComputerRetentionCheckIntervalConfiguration/config.jelly
Comment (timja) [reply core/src/main/resources/jenkins/model/GlobalComputerRetentionCheckIntervalConfiguration/config.jelly]: nothings changed in a long time there but there's plenty of legacy code, thanks for fixing

**Review by timja — APPROVED** [review]
Thanks, approving but I'm not terribly confident in this area but given your feedback on how its been running that has helped me give it a +1

**Review by StefanSpieker — APPROVED** [review]

**Comment by timja** [comment]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Comment by das7pad** [comment]
@timja Thank you for your help with getting this PR shipped!

**Comment by PiersRBME** [comment]
Hi - should I expect to find this option configurable via configuration-as-code?

**Comment by timja** [comment]
Looks like it isn't compatible, missed during review. Please file an issue.

**Pull request merged**
